src/Compression/CompressionCodecFSST.cpp

```diff
@@ -0,0 +1,145 @@
+
+        /* Copy prerequisites to dest */
+        memcpy(dest + fsst_header_size, &rows_count, sizeof(rows_count));
+        memcpy(dest + fsst_header_size + sizeof(rows_count), len_out, sizeof(len_out));
+        memcpy(dest + fsst_header_size + sizeof(rows_count) + sizeof(len_out), len_in.data(), len_in.size() * sizeof(size_t));
+
+        /* Count data total compressed size without header */
+        UInt32 compressed_size{0};
+        for (size_t i = 0; i < rows_count; ++i)
+            compressed_size += len_out[i];
+        return static_cast<UInt32>(header_size) + compressed_size;
+    }
+
+    void doDecompressData(const char * source, UInt32 source_size, char * dest, UInt32 uncompressed_size) const override
+    {
+        UNUSED(uncompressed_size, source_size);
+        fsst_decoder_t decoder;
+        size_t fsst_header_size = fsst_import(&decoder, reinterpret_cast<unsigned char *>(const_cast<char *>(source)));
+
+        size_t rows_count;
+        memcpy(&rows_count, source + fsst_header_size, sizeof(rows_count));
+
+        size_t lens[rows_count];
+        std::vector<size_t> len_in(rows_count);
+        memcpy(lens, source + fsst_header_size + sizeof(rows_count), sizeof(lens));
+        memcpy(len_in.data(), source + fsst_header_size + sizeof(rows_count) + sizeof(lens), sizeof(size_t) * len_in.size());
+        size_t header_size{fsst_header_size + sizeof(rows_count) + sizeof(lens) + sizeof(size_t) * len_in.size()};
+
+        const char * str{source + header_size};
+        for (size_t i = 0; i < rows_count; ++i)
+        {
+            dest = writeVarUInt(len_in[i], dest);
+
+            auto decompressed_size = fsst_decompress(
+                &decoder,
+                lens[i],
+                reinterpret_cast<const unsigned char *>(str),
+                OUT_SIZE,
+                reinterpret_cast<unsigned char *>(dest));
+
+            str += lens[i];
+            dest += decompressed_size;
+        }
+    }
+
+    UInt32 getMaxCompressedDataSize(UInt32 uncompressed_size) const override { return uncompressed_size + FSST_MAXHEADER; }
+
+    bool isCompression() const override { return true; }
+    bool isGenericCompression() const override { return true; }
+
+private:
+    void splitDataByRows(
+        const unsigned char * data, std::vector<const unsigned char *> & rows, std::vector<size_t> & lens, UInt32 source_size) const
+    {
+        const unsigned char * end{data + source_size};
+
+        while (data != end)
+        {
+            UInt64 cur_len;
+            data = reinterpret_cast<const unsigned char *>(readVarUInt(cur_len, reinterpret_cast<const char *>(data), source_size));
+            lens.push_back(cur_len);
+            rows.push_back(data);
+            data += cur_len;
+        }
+    }
+};
+
+void registerCodecFSST(CompressionCodecFactory & factory)
+{
+    auto codec_builder = [&](const ASTPtr & arguments) -> CompressionCodecPtr
```

Comment:
I guess we should restrict FSST to `String` and `FixedString` columns here. Check the `registerCodec` method src/Compression/CompressionCodecGorilla.cpp (which is a floating-point-value-only codec) how to do that.

Also, we should check that no further parameters were passed into `CODEC FSST()` when defined in SQL. Please see the `registerCodec` method in src/Compression/CompressionCodecGCD.cpp.

Both cases also need to have a negative SQL test, i.e. a test case in 02973_fsst_code_test_data.sql that expect an exception:

```sql
CREATE TABLE table_fsst_codec (n String CODEC(FSST('an_additional_argument'))) ENGINE = Memory; -- { serverError ILLEGAL_SYNTAX_FOR_CODEC_TYPE }
```